test/controllers/transfer/ChooseYearsControllerTest.scala

```diff
@@ -0,0 +1,199 @@
+      bind[TransferErrorHandler].toInstance(mockErrorHandler),
+      bind[TimeService].toInstance(mockTimeService),
+      bind[AuthRetrievals].to[MockAuthenticatedAction],
+      bind[MessagesApi].toInstance(stubMessagesApi()),
+      bind[PertaxAuthAction].to[FakePertaxAuthAction]
+    )
+    .build()
+
+  def controller: ChooseYearsController =
+    app.injector.instanceOf[ChooseYearsController]
+
+  when(mockTimeService.getCurrentDate) thenReturn LocalDate.now()
+  when(mockTimeService.getCurrentTaxYear) thenReturn currentTaxYear
+  when(mockTimeService.getStartDateForTaxYear(any())).thenReturn(time.TaxYear.current.starts)
+  when(mockTransferService.getCurrentAndPreviousYearsEligibility(any(), any())).thenReturn(
+    Future.successful(
+      CurrentAndPreviousYearsEligibility(
+        currentYearAvailable = true,
+        List(TaxYear(2015)),
+        RecipientRecordData.recipientRecord.data,
+        RecipientRecordData.recipientRecord.availableTaxYears
+      )
+    )
+  )
+  val recipientNotFoundView: recipient_not_found = inject[views.html.errors.recipient_not_found]
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    reset(mockCachingService)
+  }
+
+  "chooseYears" should {
+    "return OK with pre-filled form" when {
+      "cached data is available" in {
+        val cachedData = Some("currentTaxYear")
+        when(mockCachingService.get[String](any())(any()))
+          .thenReturn(Future.successful(cachedData))
+
+        val result = controller.chooseYears()(request)
+        status(result) shouldBe OK
+      }
+    }
+    "return OK with empty form" when {
+      "no cached data is available" in {
+        when(mockCachingService.get[String](any())(any()))
+          .thenReturn(Future.successful(None))
+
+        val result = controller.chooseYears()(request)
+        status(result) mustBe OK
+      }
+    }
+
+    "redirect the user " when {
+      "there are available tax years not including current year" in {
+        when(mockTransferService.getCurrentAndPreviousYearsEligibility(any(), any())).thenReturn(
+          Future.successful(
+            CurrentAndPreviousYearsEligibility(
+              currentYearAvailable = false,
+              List(TaxYear(2015)),
+              RecipientRecordData.recipientRecord.data,
+              RecipientRecordData.recipientRecord.availableTaxYears
+            )
+          )
+        )
+        val result = controller.chooseYears()(request)
+        status(result) shouldBe SEE_OTHER
+        redirectLocation(result) shouldBe Some(controllers.transfer.routes.ApplyByPostController.applyByPost().url)
+      }
+    }
+  }
+
+  "chooseYearsAction" should {
+    "return bad request" when {
+      "an empty form is submitted" in {
+        val request = FakeRequest().withMethod("POST").withFormUrlEncodedBody("value" -> "")
+
+        when(mockCachingService.put[String](ArgumentMatchers.eq(CACHE_CHOOSE_YEARS), ArgumentMatchers.eq(""))(any(), any()))
```

Comment:
You don't need to stub the caching on unsuccessful form subs as you wouldn't cache that response